File: packages/emitter/testcases/types/mapped/MappedTypes.ts
export interface Person {
  name: string;
  age: number;
  email: string;
}

// Using built-in utility types
export type PartialPerson = Partial<Person>;
export type RequiredPerson = Required<Person>;
export type ReadonlyPerson = Readonly<Person>;

// Custom mapped type
export type Nullable<T> = {
  [P in keyof T]: T[P] | null;
};

export type NullablePerson = Nullable<Person>;

// Function using mapped types
export function updatePerson(
  person: Person,
  updates: Partial<Person>
): Person {
  return { ...person, ...updates };
}

// Function with readonly parameter
export function displayPerson(person: Readonly<Person>): string {
  return `${person.name} (${person.age})`;
}

File: packages/emitter/testcases/types/mapped/MappedTypes.cs
using Tsonic.Runtime;
using Tsonic.JSRuntime;
using System.Collections.Generic;

namespace TestCases.types.mapped
{
    public class Person
    {
        public string name { get; set; }

        public double age { get; set; }

        public string email { get; set; }
    }

            public static class MappedTypes
            {
                // type PartialPerson = Partial<Person>

                // type RequiredPerson = Required<Person>

                // type ReadonlyPerson = Readonly<Person>

                // type Nullable = dynamic

                // type NullablePerson = Nullable<Person>

                public static Person updatePerson(Person person, Partial<Person> updates)
                    {
                    return new { /* ...spread */, /* ...spread */ };
                    }

                public static string displayPerson(Readonly<Person> person)
                    {
                    return $"{person.name} ({person.age})";
                    }
            }
}